### PR TITLE
[ASL] Switch to rational model & other features

### PR DESCRIPTION
**Pull request by HadrienRenaud**
This PR implements various features:

- [x] Switch to a Integer Polynomial with Rationnal coefficients model for the expression normalization
- [x] Add tests to understand how are divisions handled by `aslref`
- [x] Removes the un-used interpreter rule `SAssignTuple`
- [x] Some other refactoring

### CODE REVIEW

File: asllib/tests/division.t/run.t [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    and integer {0}
+  [1]
+
+  $ aslref static-mod-zero.asl
+  File static-mod-zero.asl, line 3, characters 19 to 26:
+  ASL Typing error: Illegal application of operator MOD on types integer {6}
+    and integer {0}
+  [1]
+
+Unsupported divisions (caught at time-checking time):
```
Comment (jalglave) [inline asllib/tests/division.t/run.t:21]: typo: time-checking time

File: asllib/tests/division.t/run.t [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  File static-mod-neg.asl, line 3, characters 19 to 27:
+  ASL Typing error: Illegal application of operator MOD on types integer {6}
+    and integer {(- 3)}
+  [1]
+
+  $ aslref static-div-undiv.asl
+  ASL Static error: Illegal application of operator DIV for values 5 and 3.
+  [1]
+
+For completeness, those operations are runtime errors:
```
Comment (jalglave) [inline asllib/tests/division.t/run.t:45]: s/runtime/dynamic (since this is the terminology that aslref uses)

File: asllib/Interpreter.ml [hunk 3]
```diff
@@ -24,6 +24,7 @@ open AST
 open ASTUtils
 
 let fatal_from pos = Error.fatal_from pos
+let unsupported_expr e = Error.(fatal_from e (UnsupportedExpr (Dynamic, e)))
```
Comment (maranget) [inline asllib/Interpreter.ml:27]: This addition entails a warning "This open statement shadows the value from_pos...".
A quick consists in swapping both lines. 
```suggestion
let unsupported_expr e = Error.(fatal_from e (UnsupportedExpr (Dynamic, e)))
let fatal_from pos = Error.fatal_from pos
```
One can also delete the local definition of fatal_from and use the qualified from `Error.fatal_from` in the rest of the file.

File: asllib/Typing.ml [hunk 4]
```diff
@@ -29,7 +29,7 @@ module TypingRule = Instrumentation.TypingRule
 let ( |: ) = Instrumentation.TypingNoInstr.use_with
 let fatal_from = Error.fatal_from
 let undefined_identifier pos x = fatal_from pos (Error.UndefinedIdentifier x)
-let unsupported_expr e = fatal_from e (Error.UnsupportedExpr e)
+let unsupported_expr e = fatal_from e Error.(UnsupportedExpr (Static, e))
```
Comment (maranget) [inline asllib/Typing.ml:32]: To suppress two "this open statement shadows the value identifier unsupported_expr...", I am in favor of deleting this line and of using the fully qualified reference `StaticInterpreter.unsupport_expr` where appropriate, which clearly states that a "static" error is flagged.
```suggestion
```

Comment (maranget) [reply asllib/Typing.ml:32]: Notice that there are three occurences of `unsupported_expr` that need to be qualified.

**Comment by maranget** [comment]
Hi @HadrienRenaud. Once the remaining warnings are silenced, the PR looks ready for merge. Thanks.

**Pull request merged**